File: 212_word-search-ii.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Solution {
    
    class Node {
         Node[] next = new Node[26];
 
         /**
          * 只有一个单词的结尾才有值
          */
         String value;
         boolean leaf;
         public Node() {
         }
     }
 
     Node root = new Node();
     int l1, l2;
 
     public List<String> findWords(char[][] board, String[] words) {
         //构建前缀树
         // 遍历所有单词
         for (String word : words) {
             Node node = root; // 每个单词都是从根节点开始构建
             // 遍历单词字母，树结构中如果已经存在则直接跳过，不存在则创建子节点
             for (int i = 0; i < word.length(); i++) {
                 if (node.next[word.charAt(i) - 'a'] == null) {
                     node.next[word.charAt(i) - 'a'] = new Node();
                 }
                 // 指向node下一个节点
                 node = node.next[word.charAt(i) - 'a'];
             }
             node.value = word;
             node.leaf = true;
         }
 
         l1 = board.length;
         if (l1 == 0) {
             return new ArrayList<>();
         }
         l2 = board[0].length;
         if (l2 == 0) {
             return new ArrayList<>();
         }
         boolean[][] visited = new boolean[l1][l2];
         //        max = l1 > l2 ? l1 : l2;
         Set<String> results = new HashSet<>();
         for (int i = 0; i < l1; i++) {
             for (int j = 0; j < l2; j++) {
                 searchTrie(board, i, j, root, results, visited);
             }
         }
         return new ArrayList<>(results);
     }
 
     public void searchTrie(char[][] board, int i, int j, Node node, Set<String> result,
             boolean[][] visited) {
         if (i < 0 || j < 0 || i >= l1 || j >= l2) {
             return;
         }
         if (visited[i][j]) {
             return;
         }
         visited[i][j] = true;
         Node next = node.next[board[i][j] - 'a'];
         if (null == next) {
             visited[i][j] = false;
             return;
         }
         if (next.leaf) {
             result.add(next.value);
         }
         searchTrie(board, i - 1, j, next, result, visited);
         searchTrie(board, i + 1, j, next, result, visited);
         searchTrie(board, i, j - 1, next, result, visited);
         searchTrie(board, i, j + 1, next, result, visited);
         visited[i][j] = false;
 
     }
 }

File: 212_word-search-ii.py
'''
1 把words中的word存入前缀树
2 对board进行深度优先搜索
dfs三个核心部分：
1 新的字符不在搜索范围内，退出
2 新的字符在搜索范围内，且该字符与之前的字符串为words中的一个word,加入结果集，并将word结束标志置0，放置重复搜索
3 按深度优先递归搜索
'''
class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        def dfs(i, j, t, s):
            ch = board[i][j]
            if ch not in t:
                return
            t = t[ch]
            if "end" in t and t["end"] == 1:
                res.append(s+ch)
                t["end"] = 0
            board[i][j] = "#"
            if i + 1 < m and board[i+1][j] != "#":
                dfs(i+1, j, t, s+ch)
            if i - 1 >= 0 and board[i-1][j] != "#":
                dfs(i-1, j, t, s+ch)
            if j + 1 < n and board[i][j+1] != "#":
                dfs(i, j+1, t, s+ch)
            if j - 1 >= 0 and board[i][j-1] != "#":
                dfs(i, j-1, t, s+ch)
            board[i][j] = ch
        
        # 将word存入前缀树    
        trie = {}
        for word in words:
            t = trie
            for ch in word:
                if ch not in t:
                    t[ch] = {}
                t = t[ch]
            t["end"] = 1
        
        m = len(board)
        n = len(board[0])
        res = []
        # 对board进行深度优先搜索
        for i in range(m):
            for j in range(n):
                dfs(i, j, trie, "")
        return res